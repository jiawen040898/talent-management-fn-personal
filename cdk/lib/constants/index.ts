export * from './resource-tag.enum';
