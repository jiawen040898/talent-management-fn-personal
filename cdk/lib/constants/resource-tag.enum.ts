export enum ResourceTag {
    SNS = 'AWS::SNS::Topic',
    SQS = 'AWS::SQS::Queue',
    IAM = 'AWS::IAM::Role',
    LAMBDA = 'AWS::Lambda::Function',
    S3 = 'AWS::S3::Bucket',
    SSM = 'AWS::SSM::Parameter',
    STATE_MACHINE = 'AWS::StepFunctions::StateMachine',
}
