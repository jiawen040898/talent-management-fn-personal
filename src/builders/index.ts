export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
