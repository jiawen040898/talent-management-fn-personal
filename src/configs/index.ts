export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
