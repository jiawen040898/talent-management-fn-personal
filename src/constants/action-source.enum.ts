export enum ActionSource {
    LEARNER = 'learner',
    MANAGER = 'manager',
    ADMIN = 'admin',
}
