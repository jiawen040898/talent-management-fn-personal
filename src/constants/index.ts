export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
