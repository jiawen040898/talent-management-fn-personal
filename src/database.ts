import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
