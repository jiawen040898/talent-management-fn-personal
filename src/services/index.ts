export * from './cognitive-score.service';
export * from './company.service';
export * from './domain-event-router.service';
export * from './employee-daily-reminder.service';
export * from './employee-management.service';
export * from './employee-progress-report.service';
export * from './employee-score.service';
export * from './feedback-cycle-action.service';
export * from './feedback-cycle-dashboard-provision.service';
export * from './feedback-cycle-reminder.service';
export * from './feedback-cycle-reviewee-recommendation.service';
export * from './fit-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './recommendation.service';
