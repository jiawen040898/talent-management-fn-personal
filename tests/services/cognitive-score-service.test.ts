import { DataSource } from 'typeorm';

import { EmployeeScoreType } from '../../src/constants';
import * as database from '../../src/database';
import { CognitiveScoreService } from '../../src/services';
import { getTestDataSourceAndAddData, testUtil } from '../setup';
import { cognitiveScoreTestData } from './fixtures/cognitive-score-service.test-data';
import { roleFitRecipeTestData } from './fixtures/role-fit-recipe.test-data';

describe('CognitiveScoreService', () => {
    let dataSource: DataSource;
    const spygetDataSource = jest.spyOn(database, 'getDataSource');

    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            cognitiveScoreTestData.entitiesToBeAdded,
        );
        spygetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    afterAll(() => {
        jest.clearAllMocks();
    });

    describe('processReasoningNumeric', () => {
        it('should return reasoning numeric score', () => {
            // Act
            const result = CognitiveScoreService.processReasoningNumeric(
                cognitiveScoreTestData.reasoningNumericAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return empty outcome for invalid assessment type', async () => {
            // Act
            const result = CognitiveScoreService.processReasoningNumeric(
                cognitiveScoreTestData.reasoningVerbalAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processReasoningVerbal', () => {
        it('should return reasoning verbal score', () => {
            // Act
            const result = CognitiveScoreService.processReasoningVerbal(
                cognitiveScoreTestData.reasoningVerbalAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return empty outcome for invalid assessment type', async () => {
            // Act
            const result = CognitiveScoreService.processReasoningVerbal(
                cognitiveScoreTestData.reasoningLogicAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processReasoningLogical', () => {
        it('should return reasoning logic score', () => {
            // Act
            const result = CognitiveScoreService.processReasoningLogical(
                cognitiveScoreTestData.reasoningLogicAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return empty outcome for invalid assessment type', async () => {
            // Act
            const result = CognitiveScoreService.processReasoningLogical(
                cognitiveScoreTestData.reasoningNumericAssessmentResult,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processReasoningAverage', () => {
        it('should return expected reasoning average score when employee score do not have any assessment completed before', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                cognitiveScoreTestData.reasoningAverageAssessmentScores,
                testUtil.mockUuid(2),
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return expected reasoning average score when employee score has assessment completed before', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                cognitiveScoreTestData.reasoningAverageForSingleAssessmentScores,
                testUtil.mockUuid(3),
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return empty outcome for invalid participant id and empty current assessment', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                [],
                testUtil.mockUuid(5),
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('getCognitiveScores', () => {
        it('should return expected cognitive score outcome', async () => {
            // Act
            const result = await CognitiveScoreService.getCognitiveScores(
                cognitiveScoreTestData.cognitiveAssessmentResults,
                cognitiveScoreTestData.participant,
                roleFitRecipeTestData.validRoleFitScoreRecipe.id,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return empty cognitive score outcome', async () => {
            // Act
            const result = await CognitiveScoreService.getCognitiveScores(
                [],
                cognitiveScoreTestData.participant,
                roleFitRecipeTestData.validRoleFitScoreRecipe.id,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should throw error when assessment questionnaire framework type is invalid', async () => {
            // Arrange
            const cognitiveAssessmentsResultsWithInvalidFrameworkType = [
                {
                    ...cognitiveScoreTestData.reasoningNumericAssessmentResult,
                    questionnaire_framework: EmployeeScoreType.WORK_INTEREST,
                },
            ];

            // Act
            const result = async () =>
                await CognitiveScoreService.getCognitiveScores(
                    cognitiveAssessmentsResultsWithInvalidFrameworkType,
                    cognitiveScoreTestData.participant,
                    roleFitRecipeTestData.validRoleFitScoreRecipe.id,
                );

            // Assert
            await expect(result()).rejects.toThrowErrorMatchingSnapshot();
        });
    });
});
