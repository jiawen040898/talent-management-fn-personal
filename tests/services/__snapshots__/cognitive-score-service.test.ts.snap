// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CognitiveScoreService getCognitiveScores should return empty cognitive score outcome 1`] = `[]`;

exports[`CognitiveScoreService getCognitiveScores should return expected cognitive score outcome 1`] = `
[
  {
    "score": 13.043478260869565,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "numeric",
        "domain_percentile": null,
        "domain_score": 0.13043478260869565,
        "ingredient_weightage": null,
      },
    },
    "score_type": "reasoning_numeric",
  },
  {
    "score": 26.666666666666668,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "verbal",
        "domain_percentile": null,
        "domain_score": 0.26666666666666666,
        "ingredient_weightage": null,
      },
    },
    "score_type": "reasoning_verbal",
  },
  {
    "score": 27.77777777777778,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "logical",
        "domain_percentile": null,
        "domain_score": 0.2777777777777778,
        "ingredient_weightage": null,
      },
    },
    "score_type": "reasoning_logical",
  },
  {
    "score": 22.49597423510467,
    "score_dimension": 0,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "reasoning_average",
        "domain_percentile": null,
        "domain_score": 0.22495974235104668,
        "ingredient_weightage": null,
      },
    },
    "score_type": "reasoning_average",
  },
]
`;

exports[`CognitiveScoreService getCognitiveScores should throw error when assessment questionnaire framework type is invalid 1`] = `"No cognitive score type found for work_interest from assessment questionnaire framework."`;

exports[`CognitiveScoreService processReasoningAverage should return empty outcome for invalid participant id and empty current assessment 1`] = `
{
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_average",
}
`;

exports[`CognitiveScoreService processReasoningAverage should return expected reasoning average score when employee score do not have any assessment completed before 1`] = `
{
  "score": 22.49597423510467,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0.22495974235104668,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_average",
}
`;

exports[`CognitiveScoreService processReasoningAverage should return expected reasoning average score when employee score has assessment completed before 1`] = `
{
  "score": 24.38003220611916,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0.24380032206119162,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_average",
}
`;

exports[`CognitiveScoreService processReasoningLogical should return empty outcome for invalid assessment type 1`] = `
{
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": undefined,
      "domain_percentile": null,
      "domain_score": 0,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_logical",
}
`;

exports[`CognitiveScoreService processReasoningLogical should return reasoning logic score 1`] = `
{
  "score": 27.77777777777778,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "logical",
      "domain_percentile": null,
      "domain_score": 0.2777777777777778,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_logical",
}
`;

exports[`CognitiveScoreService processReasoningNumeric should return empty outcome for invalid assessment type 1`] = `
{
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": undefined,
      "domain_percentile": null,
      "domain_score": 0,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_numeric",
}
`;

exports[`CognitiveScoreService processReasoningNumeric should return reasoning numeric score 1`] = `
{
  "score": 13.043478260869565,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "numeric",
      "domain_percentile": null,
      "domain_score": 0.13043478260869565,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_numeric",
}
`;

exports[`CognitiveScoreService processReasoningVerbal should return empty outcome for invalid assessment type 1`] = `
{
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": undefined,
      "domain_percentile": null,
      "domain_score": 0,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_verbal",
}
`;

exports[`CognitiveScoreService processReasoningVerbal should return reasoning verbal score 1`] = `
{
  "score": 26.666666666666668,
  "score_dimension": 0,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "verbal",
      "domain_percentile": null,
      "domain_score": 0.26666666666666666,
      "ingredient_weightage": null,
    },
  },
  "score_type": "reasoning_verbal",
}
`;
